src/aes.rs

```diff
@@ -184,3 +206,196 @@ impl<R: Read> AesReaderValid<R> {
+            encrypted_file_header: Some(encrypted_file_header),
+        })
+    }
+
+    pub fn finish(mut self) -> io::Result<W> {
+        self.write_encrypted_file_header()?;
+
+        // Zip uses HMAC-Sha1-80, which only uses the first half of the hash
+        // see https://www.winzip.com/win/en/aes_info.html#auth-faq
+        let computed_auth_code = &self.hmac.finalize_reset().into_bytes()[0..AUTH_CODE_LENGTH];
+        self.writer.write_all(computed_auth_code)?;
+
+        Ok(self.writer)
+    }
+
+    /// The AES encryption specification requires some metadata being written at the start of the
+    /// file data section, but this can only be done once the extra data writing has been finished
+    /// so we can't do it when the writer is constructed.
+    fn write_encrypted_file_header(&mut self) -> io::Result<()> {
+        if let Some(header) = self.encrypted_file_header.take() {
+            self.writer.write_all(&header)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl<W: Write> Write for AesWriter<W> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.write_encrypted_file_header()?;
+
+        // Fill the internal buffer and encrypt it in-place.
+        self.buffer.extend_from_slice(buf);
+        self.cipher.crypt_in_place(&mut self.buffer[..]);
+
+        // Update the hmac with the encrypted data.
+        self.hmac.update(&self.buffer[..]);
+
+        // Write the encrypted buffer to the inner writer.  We need to use `write_all` here as if
+        // we only write parts of the data we can't easily reverse the keystream in the cipher
+        // implementation.
+        self.writer.write_all(&self.buffer[..])?;
+
+        // Zeroize the backing memory before clearing the buffer to prevent cleartext data from
+        // being left in memory.
+        self.buffer.zeroize();
+        self.buffer.clear();
+
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.writer.flush()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::io::{self, Read, Write};
+
+    use crate::{
+        aes::{AesReader, AesWriter},
+        types::AesMode,
+    };
+
+    /// Checks whether `AesReader` can successfully decrypt what `AesWriter` produces.
+    fn roundtrip(aes_mode: AesMode, password: &[u8], plaintext: &[u8]) -> io::Result<bool> {
+        let mut buf = io::Cursor::new(vec![]);
+        let mut read_buffer = vec![];
+
+        {
+            let mut writer = AesWriter::new(&mut buf, aes_mode, &password)?;
+            writer.write_all(plaintext)?;
+            writer.finish()?;
+        }
+
+        // Reset cursor position to the beginning.
+        buf.set_position(0);
+
+        {
+            let compressed_length = buf.get_ref().len() as u64;
+            let mut reader =
+                match AesReader::new(&mut buf, aes_mode, compressed_length).validate(&password)? {
+                    None => {
+                        return Err(io::Error::new(
```

Comment:
Use ZipError::InvalidPassword instead; it was added in https://github.com/zip-rs/zip2/commit/ece098d39332f97b04e508222d0cc44ae3c1e5c4.